### PR TITLE
[Issue #85] Adding OpenBLAS as external lib.

### PR DESCRIPTION
**Pull request by iramykytyn**
Look at https://github.com/ibayer/fastFM/issues/85. OpenBLAS is needed for linking with fastfm.a.

### CODE REVIEW

**Comment by ibayer** [comment]
Please link against OpenBlas in all Makefiles. Otherwise the tests can't tell us if the library works fine with the newly added OpenBlas files.

* https://github.com/ibayer/fastFM-core/blob/master/demo/Makefile#L5
* https://github.com/ibayer/fastFM-core/blob/master/src/Makefile#L3 The Darwin check should not be necessary any-more.
* https://github.com/ibayer/fastFM-core/blob/master/src/tests/Makefile#L10

File: demo/Makefile [hunk 1]
```diff
@@ -1,9 +1,10 @@
 P= example_als_mcmc example_sgd example_sgd_bpr
 VPATH = ../
-OBJECTS= example_als_mcmc.o example_sgd example_sgd_bpr
-CFLAGS = -std=c99 -g -Wall -O0 -I../include -L../bin -I../externals/CXSparse/Include
-LDLIBS= -lfastfm -L../externals/CXSparse/Lib -lcxsparse -lblas -lm
-CC=gcc
+OBJECTS = example_als_mcmc.o example_sgd example_sgd_bpr
+INCLUDE = -I../include -L../bin -I../externals/CXSparse/Include -I../externals/OpenBLAS
```
Comment (ibayer) [inline demo/Makefile]: Do we really need `-I../externals/OpenBLAS`?

File: demo/Makefile [hunk 2]
```diff
@@ -1,9 +1,10 @@
...
+CFLAGS = -std=c99 -g -Wall -O0 $(INCLUDE)
+LDLIBS = -L../bin -lfastfm -L../externals/CXSparse/Lib -lcxsparse -L../externals/OpenBLAS -lopenblas -lpthread -lm
```
Comment (ibayer) [inline demo/Makefile]: `-L../externals/CXSparse/Lib -lcxsparse -L../externals/OpenBLAS -lopenblas -lpthread`
This  libraries should all be included in `-lfastfm`.

**Review by ibayer — CHANGES_REQUESTED** [review]
Please remove as many includes as possible from the `demo/Makefile`. Some of them should definitely be statically compiled into `fastfm`. If this doesn't work here it also won't work if we call the lib from python.

File: demo/Makefile [hunk 3]
```diff
@@ -3,7 +3,7 @@ VPATH = ../
 OBJECTS = example_als_mcmc.o example_sgd example_sgd_bpr
 INCLUDE = -I../include -L../bin -I../externals/CXSparse/Include -I../externals/OpenBLAS
 CFLAGS = -std=c99 -g -Wall -O0 $(INCLUDE)
-LDLIBS = -L../bin -lfastfm -L../externals/CXSparse/Lib -lcxsparse -L../externals/OpenBLAS -lopenblas -lpthread -lm
+LDLIBS = -L../bin -lfastfm -lopenblas -lpthread -lm
```
Comment (ibayer) [inline demo/Makefile]: why do we stil need `-lopenblas -lpthread `?

File: src/Makefile [hunk 4]
```diff
@@ -1,6 +1,6 @@
 OBJECTS = kmath.o ffm_random.o ffm_als_mcmc.o ffm_utils.o ffm_sgd.o ffm.o
 INCLUDES = -I/include -I./src -I../externals/CXSparse/Include  -I../externals/OpenBLAS
-LDLIBS = -L../externals/CXSparse/Lib -lcxsparse -L../externals/OpenBLAS -lopenblas -lpthread -lm
+LDLIBS = -L../bin -lfastfm -lpthread -lm
```
Comment (ibayer) [inline src/Makefile]: The lib is only used in `demo/Makefile`. I think this change should be reset.

File: src/tests/Makefile [hunk 5]
```diff
@@ -3,7 +3,7 @@ VPATH = ../
 OBJECTS= ffm_als_mcmc.o ffm_sgd.o
 INCLUDE = -I.. -I../../externals/CXSparse/Include -I../../externals/OpenBLAS
 CFLAGS = `pkg-config --cflags gsl glib-2.0` -std=c99 -g -Wall -O0 $(INCLUDE)
-LDLIBS= `pkg-config --libs gsl glib-2.0` -L../../externals/CXSparse/Lib -lcxsparse
+LDLIBS= `pkg-config --libs gsl glib-2.0` -L../../bin -lfastfm
```
Comment (ibayer) [inline src/tests/Makefile:7]: same as above

Comment (iramykytyn) [reply src/tests/Makefile:7]: Here it is used for linking with some blas functions which lib consists.

Comment (ibayer) [reply src/tests/Makefile:7]: Yes I understand, but it's still better to link against openblas directly here and remove `-lfastfm`. This way I don't have to (re) build the fastfm library in order to run the tests.

File: src/Makefile (hunk 4 above)
Comment (ibayer) [reply src/Makefile]: Same as below, please link against openblas instead of  `-lfastfm`.

Comment (iramykytyn) [inline src/Makefile]: lib is needed here for proper linking in cli target

File: src/tests/Makefile (hunk 5 above)
Comment (iramykytyn) [reply src/tests/Makefile:7]: Okay, I will reimplement it now.

File: demo/Makefile (hunk 1 above)
Comment (iramykytyn) [reply demo/Makefile]: removed

File: src/tests/Makefile (hunk 5 above)
Comment (iramykytyn) [reply src/tests/Makefile:7]: done!

File: demo/Makefile (hunk 2 above)
Comment (iramykytyn) [reply demo/Makefile]: fixed

File: src/Makefile (hunk 4 above)
Comment (iramykytyn) [reply src/Makefile]: changed

File: demo/Makefile (hunk 3 above)
Comment (iramykytyn) [reply demo/Makefile]: -lopenblas was added mistakenly. Now it is removed.
Add -lthread was a solution to make it build on your test machines (on mines it is building without it).  Some open blas references were undefined without it. I need more time to investigate how exactly does it affect openblas.

**Review by ibayer — APPROVED** [review]
I still don't understand why we have to link against `-lpthread` in `demo/Makefile`.

**Pull request merged**